File: java/src/main/java/cn/az/code/year2021/feb/SimplifyPath.java
package cn.az.code.year2021.feb;

import java.util.Arrays;
import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Set;

/**
 * @author ycpang
 * @since 2021-02-05 17:03
 */
public class SimplifyPath {

    public String simplifyPath(String path) {
        Deque<String> stack = new LinkedList<>();
        Set<String> skip = new HashSet<>(Arrays.asList("..", ".", ""));
        for (String dir : path.split("/")) {
            if ("..".equals(dir) && !stack.isEmpty()) {
                stack.pop();
            } else if (!skip.contains(dir)) {
                stack.push(dir);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (String dir : stack) {
            sb.insert(0, dir).insert(0, "/");
        }
        return sb.length() == 0 ? "/" : sb.toString();
    }
}

File: src/string/simplify-path.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        arr = path.split("/")
        st = []
        for d in arr:
            if not d or d == ".":
                continue
            if d != "..":
                st.append(d)
            elif st:
                st.pop()
        r = ""
        for d in st:
            r += "/" + d
        if r:
            return r
        return "/"


def main():
    dir = "/home/"
    s = Solution()
    print(s.simplifyPath(dir))


if __name__ == "__main__":
    main()
